### PR TITLE
Adds CI for unit testing TE build

### PR DESCRIPTION
**Pull request by terrykong**
TE has some integration tests that we can run, but first wanted to add unit tests and then will address integration test in follow up PR. Couldn't create a badge like the jax tests do since we need to create a new gist for it. Can address the badge in a follow up PR if preferred

### CODE REVIEW

**Comment by yhtang** [comment]
> Can you confirm this only install the TE/JAX or does it install all TE back-end?

It's using the jax-te container that we build in this repo and only installs the JAX backend as shown here: https://github.com/NVIDIA/JAX-Toolbox/blob/b73c334aa813b9a923670fb49794034fa051e052/.github/container/install-te.sh#L70

**Comment by terrykong** [comment]
> Can you confirm this only install the TE/JAX or does it install all TE back-end?

What @yhtang said + just sanity checking:

```bash
root@adbe279260ab:/opt/transformer-engine# ls -lah /opt/transformer-engine/build/**/*.so
-rwxr-xr-x 1 root root 204M May  1 18:07 /opt/transformer-engine/build/lib.linux-x86_64-3.10/libtransformer_engine.so
-rwxr-xr-x 1 root root 221K May  1 18:07 /opt/transformer-engine/build/lib.linux-x86_64-3.10/transformer_engine_jax.cpython-310-x86_64-linux-gnu.so
```
So this looks like it was only installed w/ jax support

**Comment by terrykong** [comment]
Related issue: https://github.com/NVIDIA/TransformerEngine/issues/156

File: .github/workflows/_test_te.yaml [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+      - name: Pull JAX-TE image
+        shell: bash -x -e {0}
+        run: |
+          docker pull ${{ inputs.JAX_TE_IMAGE }}
+
+      - name: Run JAX-TE unit tests w/ docker
+        shell: bash -x {0}
+        run: |
+          docker run -v /tmp/pytest-log:/tmp/pytest-log --gpus all ${{ inputs.JAX_TE_IMAGE }} sh -c "pip install pytest pytest-reportlog cuda-python && pytest --report-log=/tmp/pytest-log/report.jsonl /opt/transformer-engine/tests/jax"
```
Comment (yhtang) [inline .github/workflows/_test_te.yaml:40]: 👍 Given the richness of the JSON output, this should make it easier now to add a markdown table of all the failed tests to the [job summary](https://github.blog/2022-05-09-supercharging-github-actions-with-job-summaries/) via a subsequent job step.

Comment (yhtang) [reply .github/workflows/_test_te.yaml:40]: I'd suggest using `docker run --name=<NAME>` to split the long command line into multiple lines for better readability.

File: .github/workflows/nightly-te-test.yaml [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+          cat > endpoint.json << EOF
+          {
+            "schemaVersion": 1,
+            "label": "V100 Unit",
+            "message": "${PASSED_TESTS}/${SKIPPED_TESTS}/${FAILED_TESTS} pass/skip/fail",
+            "color": "${BADGE_COLOR}"
+          }
+          EOF
+
+      # TODO: re-enable after secret/gist_id is created for TE badge json
```
Comment (yhtang) [inline .github/workflows/nightly-te-test.yaml:86]: We can use the same gist as a single gist can contain multiple files. We just need to rename the json to `te-unit-test-status.json`, and the JAX one can become `jax-unit-test-status.json`.

**Review by yhtang — CHANGES_REQUESTED** [review]
Great work :) I only have some thoughts on small things.

File: .github/workflows/_test_te.yaml (hunk 1 above)
Comment (terrykong) [reply .github/workflows/_test_te.yaml:40]: > 👍 Given the richness of the JSON output, this should make it easier now to add a markdown table of all the failed tests to the [job summary](https://github.blog/2022-05-09-supercharging-github-actions-with-job-summaries/) via a subsequent job step.

That's a pretty cool feature!


> I'd suggest using `docker run --name=<NAME>` to split the long command line into multiple lines for better readability.

Maybe I'm missing something, but `--name` just makes it easier to find your container when you `docker ps`. Is your suggestion just to break up the shell command for readability, so something like this?

```yaml
- name: Run JAX-TE unit tests w/ docker
        shell: docker run -v /tmp/pytest-log:/tmp/pytest-log --gpus all ${{ inputs.JAX_TE_IMAGE }} sh {0}
        run: |
          pip install pytest pytest-reportlog cuda-python
          pytest --report-log=/tmp/pytest-log/report.jsonl /opt/transformer-engine/tests/jax"
```

File: .github/workflows/nightly-te-test.yaml (hunk 2 above)
Comment (terrykong) [reply .github/workflows/nightly-te-test.yaml:86]: Nice~ Easier than I thought :) 

Anything using pytest can probably re-use this badge publishing workflow (e.g., rosetta). I can address in a followup PR

File: .github/workflows/_test_te.yaml (hunk 1 above)
Comment (terrykong) [reply .github/workflows/_test_te.yaml]: Actually on second thought, my code above probably wouldn't work b/c `{0}` would be replaced by a filepath that doesn't exist within that container, so maybe something like:

```yaml
- name: Run JAX-TE unit tests w/ docker
  shell: docker run -v /tmp/pytest-log:/tmp/pytest-log --gpus all ${{ inputs.JAX_TE_IMAGE }} bash -x -c "$(cat {0})"
  run: |
    pip install pytest pytest-reportlog cuda-python
    pytest --report-log=/tmp/pytest-log/report.jsonl /opt/transformer-engine/tests/jax
```

File: .github/workflows/_test_te.yaml [hunk 3]
```diff
@@ -0,0 +1,47 @@
+name: ~test TransformerEngine
+
+on:
+  # Manual trigger
+  workflow_dispatch:
```
[the remark this answers is not in the document]
Comment (terrykong) [reply .github/workflows/_test_te.yaml]: Done

File: .github/workflows/_test_te.yaml (hunk 1 above)
Comment (yhtang) [reply .github/workflows/_test_te.yaml]: Very cool. I didn't realize we could use the `shell:` key to solve this problem.
But I wonder if "$(`cat {0}`)" could destroy certain features in the script, such as heredoc (`EOF...EOF`) and functions. I recall that GitLab has line-ending issues when defining functions in the script section.

**Review by yhtang — APPROVED** [review]

**Pull request merged**